File: modules/apps/portal-search/portal-search-web/test/js/facet_util.js
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

describe('liferay-search-facet-util', () => {
	beforeEach(done => {
		require('../../src/main/resources/META-INF/resources/js/facet_util');

		AUI().use(['liferay-search-facet-util'], () => done());
	});

	describe('removeURLParameters()', () => {
		it('removes the parameter whose name is the given key', () => {
			const parameters = ['modified=last-24-hours', 'q=test'];

			const newParameters = Liferay.Search.FacetUtil.removeURLParameters(
				'modified',
				parameters
			);

			expect(newParameters).toEqual(['q=test']);
		});

		it('does not remove parameters not matching the given key', () => {
			const parameters = [
				'modifiedFrom=2018-01-01',
				'modifiedTo=2018-01-31',
				'q=test'
			];

			const newParameters = Liferay.Search.FacetUtil.removeURLParameters(
				'modified',
				parameters
			);

			expect(newParameters).toEqual([
				'modifiedFrom=2018-01-01',
				'modifiedTo=2018-01-31',
				'q=test'
			]);
		});

		it('removes given parameter', () => {
			const parameters = Liferay.Search.FacetUtil.removeURLParameters(
				'key',
				['key=sel1', 'key=sel2']
			);

			expect(parameters).toEqual([]);
		});

		it('preserves other parameters', () => {
			const parameters = Liferay.Search.FacetUtil.removeURLParameters(
				'key1',
				['key1=sel1', 'key2=sel2']
			);

			expect(parameters).toEqual(['key2=sel2']);
		});

		it('preserves key-only parameters', () => {
			const parameters = Liferay.Search.FacetUtil.removeURLParameters(
				'key',
				['checked', 'key=value']
			);

			expect(parameters).toEqual(['checked']);
		});

		it('removes key-only parameters', () => {
			const parameters = Liferay.Search.FacetUtil.removeURLParameters(
				'checked',
				['checked', 'key=value']
			);

			expect(parameters).toEqual(['key=value']);
		});
	});

	describe('setURLParameter()', () => {
		it('adds a missing parameter', () => {
			const url = Liferay.Search.FacetUtil.setURLParameter(
				'http://example.com/',
				'q',
				'test'
			);

			expect(url).toBe('http://example.com/?q=test');
		});

		it('updates an existing parameter', () => {
			const url = Liferay.Search.FacetUtil.setURLParameter(
				'http://example.com/?q=example',
				'q',
				'test'
			);

			expect(url).toBe('http://example.com/?q=test');
		});

		it('adds a missing parameter with path', () => {
			const url = Liferay.Search.FacetUtil.setURLParameter(
				'http://example.com/path',
				'q',
				'test'
			);

			expect(url).toBe('http://example.com/path?q=test');
		});
	});

	describe('setURLParameters()', () => {
		it('adds new selections', () => {
			const parameters = Liferay.Search.FacetUtil.setURLParameters(
				'key',
				['sel1', 'sel2'],
				[]
			);

			expect(parameters).toEqual(['key=sel1', 'key=sel2']);
		});

		it('removes old selections', () => {
			const parameters = Liferay.Search.FacetUtil.setURLParameters(
				'key',
				['sel2', 'sel3'],
				['key=sel1']
			);

			expect(parameters).toEqual(['key=sel2', 'key=sel3']);
		});

		it('preserves other selections', () => {
			const parameters = Liferay.Search.FacetUtil.setURLParameters(
				'key1',
				['sel1'],
				['key2=sel2']
			);

			expect(parameters).toEqual(['key2=sel2', 'key1=sel1']);
		});
	});

	describe('.updateQueryString()', () => {
		it('removes old selections', () => {
			const queryString = Liferay.Search.FacetUtil.updateQueryString(
				'key',
				['sel2', 'sel3'],
				'?key=sel1'
			);

			expect(queryString).toEqual('?key=sel2&key=sel3');
		});

		it('adds new selections', () => {
			const queryString = Liferay.Search.FacetUtil.updateQueryString(
				'key1',
				['sel1'],
				'?key2=sel2'
			);

			expect(queryString).toEqual('?key2=sel2&key1=sel1');
		});

		it('accepts query string without question mark', () => {
			const queryString = Liferay.Search.FacetUtil.updateQueryString(
				'key1',
				['sel1'],
				'key2=sel2'
			);

			expect(queryString).toEqual('key2=sel2&key1=sel1');
		});

		it('does not prefix with ampersand', () => {
			const queryString = Liferay.Search.FacetUtil.updateQueryString(
				'key',
				['sel1', 'sel2'],
				'?'
			);

			expect(queryString).toEqual('?key=sel1&key=sel2');
		});
	});
});

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/aggregation/FacetUtil.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.portal.vulcan.aggregation;

import com.liferay.portal.kernel.search.facet.collector.FacetCollector;
import com.liferay.portal.kernel.search.facet.collector.TermCollector;
import com.liferay.portal.kernel.search.facet.config.FacetConfiguration;
import com.liferay.portal.kernel.util.GetterUtil;
import com.liferay.portal.search.aggregation.AggregationResult;
import com.liferay.portal.search.aggregation.HierarchicalAggregationResult;
import com.liferay.portal.search.aggregation.bucket.BucketAggregationResult;
import com.liferay.portal.vulcan.util.TransformUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * @author Javier Gamarra
 */
public class FacetUtil {

	public static Facet toFacet(
		BucketAggregationResult bucketAggregationResult) {

		return new Facet(
			bucketAggregationResult.getName(),
			TransformUtil.transform(
				bucketAggregationResult.getBuckets(),
				bucket -> new Facet.FacetValue(
					GetterUtil.getInteger(bucket.getDocCount()),
					bucket.getKey())));
	}

	public static Facet toFacet(
		com.liferay.portal.kernel.search.facet.Facet facet) {

		FacetCollector facetCollector = facet.getFacetCollector();

		List<TermCollector> termCollectors = facetCollector.getTermCollectors();

		if (termCollectors.isEmpty()) {
			return null;
		}

		return new Facet(
			_getFacetCriteria(facet),
			TransformUtil.transform(
				termCollectors,
				termCollector -> new Facet.FacetValue(
					termCollector.getFrequency(), termCollector.getTerm())));
	}

	public static List<Facet> toFacets(AggregationResult aggregationResult) {
		List<Facet> facets = new ArrayList<>();

		if (aggregationResult instanceof BucketAggregationResult) {
			facets.add(toFacet((BucketAggregationResult)aggregationResult));
		}
		else if (aggregationResult instanceof HierarchicalAggregationResult) {
			facets.addAll(
				toFacets((HierarchicalAggregationResult)aggregationResult));
		}

		return facets;
	}

	public static List<Facet> toFacets(
		HierarchicalAggregationResult hierarchicalAggregationResult) {

		List<Facet> facets = new ArrayList<>();

		Map<String, AggregationResult> aggregationResults =
			hierarchicalAggregationResult.getChildrenAggregationResultsMap();

		for (AggregationResult aggregationResult :
				aggregationResults.values()) {

			facets.addAll(toFacets(aggregationResult));
		}

		return facets;
	}

	private static String _getFacetCriteria(
		com.liferay.portal.kernel.search.facet.Facet facet) {

		FacetConfiguration facetConfiguration = facet.getFacetConfiguration();

		if (facetConfiguration.getLabel() != null) {
			return facetConfiguration.getLabel();
		}

		return facet.getFieldName();
	}

}